### PR TITLE
Rip out Loopback. Use json schema for API.

### PR DESCRIPTION
**Pull request by awong-dev**
This removes all off loopback in favor of directly coding the rest
endpoint in express.

The impetus was loopback-connector-soap abstracted away the ability to
handle the certificate authority making the TLS setup harder. Also, the
connector's function was to bridge the internal SOAP API back outwards
which is functionality we don't want.

The new code is more straight forward.

Furthermore, the REST API has been changed so that submissions happen via
a POST to

  /api/hca/v1/applications

Status checks will eventually be exposed via a GET on

  /api/hca/v1/application/[:id]

which makes more sense based on REST semantics.

The applications schema is now defined in src/common/schema/application.js
using jsonschema. This gives a harder split between frontend model and
rest model allowing for the REST model to have stricter type checking that
makes the mapping code easier.

Code was tested against the characterization test. Actual unittests should be
added to replace the black-box characterization test.

As of right now, pre-prod connectivity is slighty broken because the
wsdl download fails the strict SSL checks. Debugging the CA problems is
the next step.

### CODE REVIEW

File: src/common/veteran.js [hunk 1]
```diff
@@ -642,4 +643,68 @@ const completeVeteran = {
...
+      case 'provideSupportLastYear':
+      case 'receivesVaPension':
+      case 'provideFinancialInfo':
+      case 'understandsFinancialDisclosure':
+      case 'sameAddress':
+      case 'cohabitedLastYear':
+      case 'isCoveredByHealthInsurance':
+      case 'isMedicaidEligible':
+      case 'isEnrolledMedicarePartA':
+        return value.value === 'Y';
```
Comment (robbiethegeek) [inline src/common/veteran.js]: by default are the values N if not submitted or will these just become undefined?

File: src/common/veteran.js [hunk 2]
```diff
@@ -642,4 +643,68 @@ const completeVeteran = {
...
+      case 'grossIncome':
+      case 'netIncome':
+      case 'otherIncome':
+      case 'deductibleEducationExpenses':
+      case 'deductibleFuneralExpenses':
+      case 'deductibleMedicalExpenses':
+        return Number(value.value);
+
+      default:
+        // fall through.
```
Comment (robbiethegeek) [inline src/common/veteran.js]: may trigger an eslint error

File: src/common/veteran.js [hunk 3]
```diff
@@ -642,4 +643,68 @@ const completeVeteran = {
...
+    }
+
+    // Turn date fields into ISO8601 dates.
+    if (value.day !== undefined && value.month !== undefined && value.year !== undefined) {
```
Comment (robbiethegeek) [inline src/common/veteran.js]: we should use moment if it is still in the file or readd : http://momentjs.com/docs/#/displaying/as-iso-string/

File: src/server.js [hunk 4]
```diff
@@ -41,17 +41,17 @@ function makeServer() {
...
+  return app;
 }
 
-const server = makeServer();
-const api = loopback();
-voaRest.attach(api);
-server.use('/', api);
+const app = makeApp();
+app.use(bodyParser.json());
+app.post(`${config.apiRoot}/applications`, api.submitApplication);
```
Comment (robbiethegeek) [inline src/server.js]: why both routes to the same method?

File: src/server/api.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  config.soap.wsdl,
+  {
+    security: securityArtifacts,
+    wsdl_options: securityArtifacts === null ? null : { // eslint-disable-line
+      rejectUnauthorized: false,
+      strictSSL: false,
+      requestCert: true
+    }
+  },
+  (err, client) => { voaService = client; });
```
Comment (robbiethegeek) [inline src/server/api.js]: we should deal with the error if we try to spin up and the server is offline it will blow up

File: src/server/api.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+// TODO(awong): Remove config.url.
+
+function submitApplication(req, res) {
+  const contentType = req.get('Content-Type');
+  if (contentType !== 'application/json') {
+    res.status(400).json({ error: `Expects application/json content-type. Got "${contentType}"` });
+    return;
+  }
+
+  console.log(JSON.stringify(req.body, null, 2));
```
Comment (robbiethegeek) [inline src/server/api.js]: add todo this via a logger.

File: src/server/enrollment-system.js [hunk 7]
```diff
@@ -263,12 +98,12 @@ function spanishHispanicToSDSCode(isSpanishHispanicLatino) {
...
-  if (veteran.isAsian) races.push({ race: '2028-9' });
-  if (veteran.isBlackOrAfricanAmerican) races.push({ race: '2054-5' });
-  if (veteran.isNativeHawaiianOrOtherPacificIslander) races.push({ race: '2076-8' });
-  if (veteran.isWhite) races.push({ race: '2106-3' });
-  return races;
+  if (veteran.isAmericanIndianOrAlaskanNative) races.push('1002-5');
+  if (veteran.isAsian) races.push('2028-9');
+  if (veteran.isBlackOrAfricanAmerican) races.push('2054-5');
+  if (veteran.isNativeHawaiianOrOtherPacificIslander) races.push('2076-8');
+  if (veteran.isWhite) races.push('2106-3');
```
Comment (robbiethegeek) [inline src/server/enrollment-system.js]: is this the valid JSON for sequential elements?  Not sure of the format

**Comment by robbiethegeek** [comment]
This looks good I don't have any real practical JSON schema experience, I am going to pass to @alexose for another set of eyes.

File: src/common/veteran.js (hunk 2 above)
Comment (awong-dev) [inline src/common/veteran.js]: Tested locally. It doesn't. \o/

File: src/common/veteran.js (hunk 3 above)
Comment (awong-dev) [inline src/common/veteran.js]: I will add a TODO but I had problems with timezones (as described on slack) such that I didn't want to transition through the native Date object.

File: src/common/schema/application.js [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+      type: 'boolean'
+    },
+    radiumTreatments: {
+      type: 'boolean'
+    },
+    campLejeune: {
+      type: 'boolean'
+    }
+  }
+};
```
Comment (alexose) [inline src/common/schema/application.js]: Without verifying that this is 100% schematically correct, I will say that this is awesome a big improvement 👍

File: src/common/veteran.js (hunk 3 above)
Comment (alexose) [inline src/common/veteran.js]: Maybe a separate PR to transition these things to moment?

File: src/server.js (hunk 4 above)
Comment (awong-dev) [inline src/server.js]: oops. that was a copy-pasta mistake.  The singular version is removed.

File: src/server/api.js (hunk 5 above)
Comment (awong-dev) [inline src/server/api.js]: Yeah...I don't know how to do in node.js. Like what's the format of the `err` parameter? How do I reconnect? etc. etc.

Adding TODO now...

File: src/server/api.js (hunk 6 above)
Comment (awong-dev) [inline src/server/api.js]: I'll just delete.

File: src/server.js (hunk 4 above)
Comment (alexose) [inline src/server.js]: Backwards compatibility with loopback's shady plural conventions, I assume

File: src/server/enrollment-system.js (hunk 7 above)
Comment (awong-dev) [inline src/server/enrollment-system.js]: yup.

I couldn't find a spec or anything for the format, but it generates sensible looking xml when I capture it.

File: src/server/api.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
+'use strict'; // eslint-disable-line
+
+const fs = require('fs');
+const soap = require('soap');
+const debug = require('debug')('hca:api');
+const ajv = require('ajv');
```
Comment (alexose) [inline src/server/api.js]: `ajv` looks cool.  Approve.

**Comment by alexose** [comment]
This looks great, @awong-dev! 👍

File: src/common/veteran.js (hunk 1 above)
Comment (awong-dev) [inline src/common/veteran.js]: They will just become undefined. Which I think defaulting to false is okay?

File: src/common/veteran.js (hunk 3 above)
Comment (awong-dev) [inline src/common/veteran.js]: I remain unconvinced that moment is a good idea for this particular action.

**Pull request merged**